ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/Node.hs

```diff
@@ -0,0 +1,346 @@
+import           Control.Tracer (nullTracer)
+
+import qualified Data.ByteString.Lazy as BL
+import           Data.IP (toIPv4, IP (..), toIPv6)
+import           Data.List ((\\))
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Void (Void)
+import           System.Random (StdGen, mkStdGen)
+
+import           Network.DNS (Domain)
+
+import           Ouroboros.Network.Driver.Limits
+                   (ProtocolSizeLimits(..), ProtocolTimeLimits (..))
+import           Ouroboros.Network.Mux (MiniProtocolLimits(..))
+import           Ouroboros.Network.NodeToNode.Version (DiffusionMode (..))
+import           Ouroboros.Network.Protocol.ChainSync.Codec
+                   (byteLimitsChainSync, timeLimitsChainSync,
+                      ChainSyncTimeout (..))
+import           Ouroboros.Network.Protocol.Handshake.Version (Accept (Accept))
+import           Ouroboros.Network.Protocol.KeepAlive.Codec
+                   (byteLimitsKeepAlive, timeLimitsKeepAlive)
+import           Ouroboros.Network.Protocol.Limits (shortWait, smallByteLimit)
+import           Ouroboros.Network.PeerSelection.Governor
+                   (PeerSelectionTargets (..))
+import           Ouroboros.Network.PeerSelection.LedgerPeers
+                   (LedgerPeersConsensusInterface (..), UseLedgerAfter (..))
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAccessPoint (..),
+                   LookupReqs (..), RelayAccessPoint (..), PortNumber)
+import           Ouroboros.Network.PeerSelection.Types (PeerAdvertise (..))
+import           Ouroboros.Network.Server.RateLimiting
+                   (AcceptedConnectionsLimit (..))
+import           Ouroboros.Network.Snocket (TestAddress (..))
+
+import           Ouroboros.Network.Testing.ConcreteBlock (Block)
+import           Ouroboros.Network.Testing.Data.Script (Script (..))
+import           Simulation.Network.Snocket
+                   ( withSnocket,
+                   BearerInfo (..) )
+
+import           Test.Ouroboros.Network.Diffusion.Node.NodeKernel (NtNAddr, randomBlockGenerationArgs, BlockGeneratorArgs)
+import           Test.Ouroboros.Network.PeerSelection.RootPeersDNS (DNSTimeout,
+                   DNSLookupDelay)
+import qualified Test.Ouroboros.Network.Diffusion.Node.NodeKernel    as Node
+import qualified Test.Ouroboros.Network.Diffusion.Node.MiniProtocols as Node
+import qualified Test.Ouroboros.Network.Diffusion.Node               as Node
+
+import           Test.QuickCheck
+                   ( Arbitrary(arbitrary),
+                     Gen,
+                     choose,
+                     chooseInt,
+                     sized,
+                     vectorOf,
+                     suchThat, oneof
+                   )
+
+-- | Diffusion Simulator Arguments
+--
+-- Contains all necessary randomly generated values needed to run diffusion in
+-- simulation.
+--
+type SimArgs =
```

Comment:
I think it it deserves a data type with proper haddocks.

Reply:
Will do!